### PR TITLE
fixes #682 add route for bulk posting things and moving their files to the correct place based on controller redirect

### LINKED ISSUE #682: Add route for bulk posting ls things and moving files to configured redirect conf ls thing file path

### PR DESCRIPTION
**Pull request by brianbolt**

### CODE REVIEW

File: modules/ServerAPI/src/server/routes/ThingServiceRoutes.coffee [hunk 1]
```diff
@@ -285,6 +287,42 @@ updateThing = (thing, testMode, callback) ->
...
+			filesToSave = fileVals.length
+
+			completeThingUpdate = (thingToUpdate)->
+				updateThing thingToUpdate, false, (updatedThing) ->
+					callback updatedThing, 200
+
+			fileSaveCompleted = (passed) ->
+				if !passed
+					callback "file move failed", 500
+				if --filesToSave == 0 then completeThingUpdate(thing)
```
Comment (bffrost) [inline modules/ServerAPI/src/server/routes/ThingServiceRoutes.coffee:303]: What is this `--filesToSave` doing? Decrementing by one?

File: modules/ServerAPI/src/server/routes/ThingServiceRoutes.coffee [hunk 2]
```diff
@@ -285,6 +287,42 @@ updateThing = (thing, testMode, callback) ->
...
+
+			if filesToSave > 0
+				prefix = serverUtilityFunctions.getPrefixFromEntityCode thing.codeName
+				for fv in fileVals
+					csUtilities.relocateEntityFile fv, prefix, thing.codeName, fileSaveCompleted
+			else
+				callback thing, 200
+
+		if response.indexOf("saveFailed") > -1
```
Comment (bffrost) [inline modules/ServerAPI/src/server/routes/ThingServiceRoutes.coffee:312]: oh, this is where the actual function calls for the callback from `bulkPostThingsInternal` start.

File: modules/ServerAPI/src/server/routes/ThingServiceRoutes.coffee (hunk 1 above)
Comment (bffrost) [reply modules/ServerAPI/src/server/routes/ThingServiceRoutes.coffee:303]: I'm really confused about where the actual changes to the fileValues happen. We seem to just be passing the `thing` that comes back from the "bulk post" and then calling "updateThing" on it. When do the file values actually change?

File: modules/ServerAPI/src/server/routes/ThingServiceRoutes.coffee [hunk 3]
```diff
@@ -285,6 +287,42 @@ updateThing = (thing, testMode, callback) ->
...
+
+			fileSaveCompleted = (passed) ->
+				if !passed
+					callback "file move failed", 500
+				if --filesToSave == 0 then completeThingUpdate(thing)
+
+			if filesToSave > 0
+				prefix = serverUtilityFunctions.getPrefixFromEntityCode thing.codeName
+				for fv in fileVals
+					csUtilities.relocateEntityFile fv, prefix, thing.codeName, fileSaveCompleted
```
Comment (bffrost) [inline modules/ServerAPI/src/server/routes/ThingServiceRoutes.coffee:308]: Is `fv` being modified by this function, and the change flowing all the way back to `thing`?

Comment (brianbolt) [reply modules/ServerAPI/src/server/routes/ThingServiceRoutes.coffee:308]: relocate entity file is modifying the thing and then calling its call back "fileSaveCompleted".  This way, the customer specific server function is responsible for doing something with the file.

**Comment by brianbolt** [comment]
I'll add comments to the code. Most of this was copied from function within anothing thing save route.

File: modules/ServerAPI/src/server/routes/ThingServiceRoutes.coffee (hunk 1 above)
Comment (brianbolt) [reply modules/ServerAPI/src/server/routes/ThingServiceRoutes.coffee:312]: I added comments to clarify but yes, it is decrementing by one.

File: modules/ServerAPI/src/server/routes/ThingServiceRoutes.coffee (hunk 3 above)
Comment (bffrost) [reply modules/ServerAPI/src/server/routes/ThingServiceRoutes.coffee:319]: I guess I am just confused that this `relocateEntityFile` function is not having the `thing` variable passed to it. How does it access the `thing`? Does it look it up by codeName?

File: modules/ServerAPI/src/server/routes/ThingServiceRoutes.coffee (hunk 1 above)
Comment (bffrost) [reply modules/ServerAPI/src/server/routes/ThingServiceRoutes.coffee:312]: Similar question here... This `fileSaveCompleted` function is only getting the true/false `passed` variable as an input. Where is the value of `thing` coming from?

File: modules/ServerAPI/src/server/routes/ThingServiceRoutes.coffee (hunk 3 above)
Comment (brianbolt) [reply modules/ServerAPI/src/server/routes/ThingServiceRoutes.coffee:325]: its getting passed just the file value `fv`. I just added some more code comments to clarify (I had to look it up to).

**Pull request merged**